src/main/java/fi/aalto/cs/apluscourses/ui/exercise/ExercisesView.java

```diff
@@ -1,26 +1,52 @@
 package fi.aalto.cs.apluscourses.ui.exercise;
 
+import static fi.aalto.cs.apluscourses.utils.PluginResourceBundle.getText;
+
 import com.intellij.openapi.application.ApplicationManager;
 import com.intellij.openapi.application.ModalityState;
 import fi.aalto.cs.apluscourses.intellij.actions.ActionUtil;
 import fi.aalto.cs.apluscourses.intellij.actions.OpenSubmissionAction;
+import fi.aalto.cs.apluscourses.presentation.MainViewModel;
 import fi.aalto.cs.apluscourses.presentation.exercise.ExercisesTreeViewModel;
 import fi.aalto.cs.apluscourses.ui.GuiObject;
 import fi.aalto.cs.apluscourses.ui.base.TreeView;
+
+import java.awt.CardLayout;
+
+import javax.swing.JLabel;
 import javax.swing.JPanel;
+import javax.swing.JScrollPane;
+import javax.swing.SwingConstants;
+
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
+
 public class ExercisesView {
   private TreeView exerciseGroupsTree;
+  private JLabel emptyText;
   private JPanel basePanel;
-
+  private JScrollPane pane;
   @GuiObject
   public JPanel toolbarContainer;
+  private JPanel cardPanel;
+  private CardLayout cl;
+  private MainViewModel mainViewModel;
+  private boolean hasCourse = false;
 
-  public ExercisesView() {
-    // See ModulesView.java
+  /**
+   * Creates an ExerciseView that uses mainViewModel to dynamically adjust its UI components.
+   */
+  public ExercisesView(MainViewModel mainViewModel) {
+    this.mainViewModel = mainViewModel;
+    mainViewModel.exercisesViewModel
+            .addValueObserver(this, ExercisesView::viewModelChanged);
```

Comment:
Please, move this to the end of the ctor. `ObservableProperty::addValueObserver` calls immediately `this::viewModelChanged`, so it is possible that value for `cl` is not yet set when it is referenced in `updateComponents` (called by `viewModelChanged`).

I must apologize that I made you move that observer registration here; I now see it's a bit controversal to have it here because it "leaks" not-yet-constructed `this` (which is a commonly known source source of anomalies). It's however sage to have it here, as long as it's the last statement of the ctor (btw, you don't have to call `updateComponents` separately here, you can remove that line). Other possibility is that it's called from the caller of this ctor, as it was previously (sorry for confusion).